Develop

upostnikova0 opened the pull request:
Реализовано:
1) Вывод общих с другом фильмов с сортировкой по их популярности.
2) Удаление фильмов и пользователей
3) Добавление режиссеров в фильмы:
   1) Вывод всех фильмов режиссёра, отсортированных по количеству лайков.
   2) Вывод всех фильмов режиссёра, отсортированных по годам.
4) Функциональность "Отзывы". Добавленные отзывы должны иметь рейтинг и несколько дополнительных характеристик:
   1) Оценка — полезно/бесполезно.
   2) Тип отзыва — негативный/положительный.
   3) Рейтинг отзыва. При создании отзыва рейтинг равен нулю. Если пользователь оценил отзыв как полезный, это увеличивает его рейтинг на 1. Если как бесполезный, то уменьшает на 1. 
Отзывы должны сортироваться по рейтингу полезности.
5) Вывод самых популярных фильмов по жанру и годам. 
6) Возможность просмотра последних событий на платформе — добавление в друзья, удаление из друзей, лайки и отзывы, которые оставили друзья пользователя. 
7) Поиск по названию фильмов и по режиссёру.
8)  Рекомендательная система для фильмов.

upostnikova0 committed:
feat: added functionality to delete films and users

upostnikova0 committed:
feat: added functionality to delete films and users

upostnikova0 committed:
feat: added functionality to delete films and users

upostnikova0 committed:
feat: added functionality to delete films and users

upostnikova0 committed:
style: rermoved unused import

upostnikova0 committed:
style: rermoved unused import

upostnikova0 committed:
style: rermoved unused import

upostnikova0 committed:
style: rermoved unused import

TeDixxx committed:
add common-films (#8)

* add common-films

* Delete InMemory and add test

---------

Co-authored-by: Dmitry <dm1try-teter1n@yandex.ru>

upostnikova0 committed:
feat: add directors

upostnikova0 committed:
feat: add directors

upostnikova0 committed:
style: removed unused import at FilmDbStorage

upostnikova0 committed:
refactor: added cascading deletion of records

upostnikova0 committed:
refactor: added cascading deletion of records

upostnikova0 committed:
refactor: added cascading deletion of records

upostnikova0 committed:
Merge branch 'add-remove-endpoint' of github.com:upostnikova0/java-filmorate into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java
#	src/test/java/ru/yandex/practicum/filmorate/storage/FilmStorageTests.java
#	src/test/java/ru/yandex/practicum/filmorate/storage/UserStorageTests.java

upostnikova0 committed:
style: rename test package

upostnikova0 committed:
Merge pull request #7 from upostnikova0/add-remove-endpoint

add-remove-endpoint

upostnikova0 committed:
Merge branch 'develop' into add-director

upostnikova0 committed:
style: removed unused import at class FilmDbStorage

upostnikova0 committed:
style: rename test package

upostnikova0 committed:
Merge pull request #9 from upostnikova0/add-director

feat: add directors

JRRT93 committed:
feat: отзывы

IvanTurk100 committed:
Merge pull request #15 from upostnikova0/add-reviews

feat: отзывы

IvanTurk100 committed:
feat: добавил вывод самых популярных фильмов по жанру и годам

upostnikova0 committed:
Merge pull request #17 from upostnikova0/add-most-populars

add most populars

upostnikova0 committed:
feat: add-feed

upostnikova0 committed:
style: added separator line

upostnikova0 committed:
style: removed unused import at LikesDbStorage

upostnikova0 committed:
Merge pull request #18 from upostnikova0/add-feed

feat: add-feed

upostnikova0 committed:
feat: add search

upostnikova0 committed:
Merge pull request #20 from upostnikova0/add-search

feat: add search

upostnikova0 committed:
fix: sql query

upostnikova0 committed:
style: schema.sql

upostnikova0 committed:
style: add .gitignore

upostnikova0 committed:
style: removed checkstyle.xml

upostnikova0 committed:
revert: checkstyle.xml

upostnikova0 committed:
revert checkstyle

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -104,76 +168,168 @@ public Film update(Film film) {
+
+    public Collection<Film> search(String query, String by) {
+        List<Film> searchFilms = new ArrayList<>();
+        List<Film> allFilms = new ArrayList<>(filmStorage.findAll());
```
загружаем много лишних данных, фильтрацию лучшне сделать через запросы, 
т.е.  нужно написать три запроса в зависимости от условий поиска: title, director, title + director

в реальных проектах БД не всегда находится на одной машине с приложением, а иногда может и в другом городе. Выкачивать всех данные, чтобы выбрать несколько фильмов будет избыточной нагрузкой.

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/storage/event/EventDbStorage.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public Collection<Event> findAll(long userId) {
+        String sql = "SELECT * FROM EVENTS WHERE USER_ID = ?";
```
по т.з просили "просмотра последних событий на платформе", т.е. в запросе указываем сортировку 
`ORDER BY EVENT_ID`

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDbStorage.java:
```diff
@@ -0,0 +1,152 @@
+
+    @Override
+    public void delete(long reviewId) {
+        String sqlQuery = "DELETE FROM reviews WHERE review_id = ?";
```
вместо удаление записи из таблицы, лучше добавить поле DELETED
```
create table if not exists REVIEWS (
...
deleted     bool default false
);
```

и при удалении проставлять true.
Для FILMS и USERS лучше сделать аналогично.
Оставлять записи в БД может быть полезно, если удалили по ошибки и потом просят возвратить запись.

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,41 @@
+    @NotNull
+    private Long filmId;
+    private Integer useful;
+    private Set<Long> likes;
```
likes и dislikes  не просили в Review, судя по JSON примеру из Т.З.
если убрать эти поля, то код существенно упростится

IvanTurk100 committed:
fix: all successful postman tests

IvanTurk100 committed:
Merge pull request #23 from upostnikova0/add-most-populars

Add most populars #2

upostnikova0 committed:
Add recommendations (#24)

* feat: added recommendations and tests

upostnikova0 committed:
fix: added like if exist (#25)

upostnikova0 committed:
Add remove endpoint (#26)

refactor: added soft delete users, films, reviews

TeDixxx committed:
Add reviews

*refactor: Review

---------

Co-authored-by: Dmitry <dm1try-teter1n@yandex.ru>

Vitalya1991 committed:
Add search (#28)

* add-search

---------

Co-authored-by: francesy <upostnikova0@yandex.ru>
Co-authored-by: upostnikova0 <110557030+upostnikova0@users.noreply.github.com>

upostnikova0 committed:
add-search (#30)

* fix: return found films

---------

Co-authored-by: Vitalya1991 <convinceall@mail.ru>

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -51,35 +84,64 @@ public Film add(Film film) {
 
-        if (!allGenres.isEmpty()) {
+    private Collection<Film> getFilmsWithAllFields(List<Film> allFilms, List<Map<Long, Genre>> allGenres, List<Map<Long, Director>> allDirectors) {
+        if (allGenres != null) {
```
1) цикл лучше сделать по allFilms
2) `List<Map<Long, Genre>> allGenres` лушче заменить на `Map<Long, List<Genre>> allGenres `
тогда за один  поиск по filmId пожно получить список Genre для фильма

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDbStorage.java:
```diff
@@ -0,0 +1,162 @@
+            reviews = jdbcTemplate.query(sqlQuery, rowMapper, count);
+        }
+
+//        sqlQuery = "SELECT * FROM reviews_likes";
```
закомментаренный код лучше убрать

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDbStorage.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public void addOrDeleteLikeOrDislike(long reviewId, long userId, String likeOrDislike, String requestMethod) {
```
лучше разделить на отдельные методы, не очень читабельно получилось

JRRT93 committed:
Add reviews (#31)

refactor: метод deleteOrAddLikeOrDislike() разделен на отдельные методы для улучшения читаемости кода во всех слоях
fix: удалён весь неактуальный код для реализации функционала отзывы. Починены юнит-тесты
refactor: изменены методы findAll() для жанров и режиссеров, изменен метод формирования фильмов со всеми полями в FilmService, в прочих методах исправлены типы возвращаемых данных в соответствии с внесенными изменениями

---------

Co-authored-by: Dmitry <dm1try-teter1n@yandex.ru>

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,65 @@
+        if (likeOrDislike.equals("like")) {
+            reviewService.addLike(reviewId, userId);
+        }
+        if (likeOrDislike.equals("dislike")) {
```
так как мы можем зайти только в один if, то для второго условия лучше поставить 
`else if (likeOrDislike.equals("dislike"))`. Это особенно актуально, если подряд идёт много условий.

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -6,41 +6,74 @@
+
+    public Collection<Film> getSortedFilmsByDirectorId(long directorId, String filter) {
+        directorService.getDirector(directorId);
+        Map<Long, List<Genre>> allGenres = filmGenresStorage.findAll();
```
такой вариант уже будет лучше работать, но жанры можно грузить только для выбранных фильмов
для этого надо добавить в `FilmGenresStorage` метод для загрузки жанров для фильмов
```
    @Override
    public Map<Long, List<Genre>> load(List<Film> films) {
        String inSql = String.join(",", Collections.nCopies(films.size(), "?"));//склеиваем знаки вопроса через запятую
        String sql = "select * from GENRES g, FILM_GENRES fg where fg.GENRE_ID = g.GENRE_ID AND fg.FILM_ID in (" + inSql + ")";
        Map<Long, List<Genre>> result = new LinkedHashMap<>();
        jdbcTemplate.query(sql, (rs, rowNum) -> {
            result.putIfAbsent(rs.getLong("film_id"), new ArrayList<>());
            result.get(rs.getLong("film_id")).add(new Genre(rs.getInt("genre_id"),
                    rs.getString("genre_name")));
            return result;
        }, films.stream().map(Film::getId).toArray());
        return result;
    }
```

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -51,35 +84,54 @@ public Film add(Film film) {
+            }
+            if (allDirectors != null) {
+                Optional<List<Director>> directors = Optional.ofNullable(allDirectors.get(film.getId()));
+                film.setDirectors(directors.orElse(new ArrayList<>()));
```
если мы не планируем добавлять объекты в лист, то лучше заменить `new ArrayList<>()` на `Collections.emptyList()`
чтобы не создавать лишние объекты

Dmitry-E commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -83,6 +128,26 @@ public Collection<User> getCommonFriends(long id, long friendId) {
+
+        if (allGenres != null) {
+            for (Film film : recommendFilms) {
+                if (allGenres.containsKey(film.getId())) {
```
лучше не делать лишний поиск `containsKey` и `get`. Достаточно сделать get
```
                var genres = allGenres.get(film.getId());
                if (genres != null) {
                    film.setGenres(genres);
                }
```

Dmitry-E reviewed (commented):
Получилось хорошо. Написал, что ещё можно улучшить.
